File: android/app/src/main/java/com/example/p2pcall/signaling/SignalingClient.kt
package com.example.p2pcall.signaling

import android.util.Log
import io.socket.client.Ack
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject

/**
 * Signaling layer for Android. Thin wrapper over the Socket.IO client that
 * mirrors the documented protocol. Knows nothing about WebRTC; it only relays
 * messages and surfaces lifecycle events through [Listener].
 */
class SignalingClient(
    private val url: String,
    private val listener: Listener,
) {
    interface Listener {
        fun onConnected()
        fun onDisconnected()
        fun onReconnected()
        fun onPeerJoined(peerId: String)
        fun onPeerLeft(peerId: String)
        fun onSignal(from: String, data: JSONObject)
    }

    private var socket: Socket? = null

    val id: String? get() = socket?.id()

    fun connect() {
        if (socket != null) return
        val opts = IO.Options().apply {
            transports = arrayOf("websocket")
            reconnection = true
            reconnectionDelay = 1000
            reconnectionDelayMax = 5000
        }
        val s = IO.socket(url, opts)

        s.on(Socket.EVENT_CONNECT) { listener.onConnected() }
        s.on(Socket.EVENT_DISCONNECT) { listener.onDisconnected() }
        s.io().on("reconnect") { listener.onReconnected() }

        s.on("peer-joined") { args ->
            (args.firstOrNull() as? JSONObject)?.let {
                listener.onPeerJoined(it.optString("peerId"))
            }
        }
        s.on("peer-left") { args ->
            (args.firstOrNull() as? JSONObject)?.let {
                listener.onPeerLeft(it.optString("peerId"))
            }
        }
        s.on("signal") { args ->
            (args.firstOrNull() as? JSONObject)?.let {
                val from = it.optString("from")
                val data = it.optJSONObject("data") ?: return@let
                listener.onSignal(from, data)
            }
        }

        socket = s
        s.connect()
    }

    /**
     * Join a room. The ack carries the existing peers (empty when first in).
     * @param onAck invoked with (ok, peers, reason)
     */
    fun joinRoom(roomId: String, onAck: (ok: Boolean, peers: List<String>, reason: String?) -> Unit) {
        val payload = JSONObject().put("roomId", roomId)
        socket?.emit("join-room", arrayOf(payload), Ack { response ->
            val obj = response.firstOrNull() as? JSONObject
            if (obj == null) {
                onAck(false, emptyList(), "no-response")
                return@Ack
            }
            if (obj.optBoolean("ok", false)) {
                val arr = obj.optJSONArray("peers")
                val peers = buildList {
                    if (arr != null) for (i in 0 until arr.length()) add(arr.getString(i))
                }
                onAck(true, peers, null)
            } else {
                onAck(false, emptyList(), obj.optString("reason", "unknown"))
            }
        })
    }

    fun sendSignal(to: String, data: JSONObject) {
        val payload = JSONObject().put("to", to).put("data", data)
        socket?.emit("signal", payload)
    }

    fun leaveRoom(roomId: String) {
        socket?.emit("leave-room", JSONObject().put("roomId", roomId))
    }

    fun disconnect() {
        socket?.let {
            it.off()
            it.io().off()
            it.disconnect()
            it.close()
        }
        socket = null
        Log.d(TAG, "Signaling disconnected")
    }

    private companion object {
        const val TAG = "SignalingClient"
    }
}

File: ios/P2PCall/Signaling/SignalingClient.swift
import Foundation
import SocketIO

/// Signaling layer for iOS. Thin wrapper over the Socket.IO client that mirrors
/// the documented protocol (see repo README). Knows nothing about WebRTC; it
/// only relays messages and surfaces lifecycle events through closures, the
/// counterpart to Android's `SignalingClient`.
final class SignalingClient {

    struct Listener {
        var onConnected: () -> Void = {}
        var onDisconnected: () -> Void = {}
        var onReconnected: () -> Void = {}
        var onPeerJoined: (_ peerId: String) -> Void = { _ in }
        var onPeerLeft: (_ peerId: String) -> Void = { _ in }
        var onSignal: (_ from: String, _ data: [String: Any]) -> Void = { _, _ in }
    }

    private let url: String
    private let listener: Listener

    private var manager: SocketManager?
    private var socket: SocketIOClient?
    /// First `connect` is a fresh join; subsequent ones are reconnects (mirrors
    /// the Android split between EVENT_CONNECT and the io "reconnect" event).
    private var hasConnectedOnce = false

    init(url: String, listener: Listener) {
        self.url = url
        self.listener = listener
    }

    func connect() {
        guard socket == nil, let parsed = URL(string: url) else { return }
        let mgr = SocketManager(socketURL: parsed, config: [
            .forceWebsockets(true),
            .reconnects(true),
            .reconnectWait(1),
            .reconnectWaitMax(5),
            .compress,
        ])
        let s = mgr.defaultSocket

        s.on(clientEvent: .connect) { [weak self] _, _ in
            guard let self = self else { return }
            if self.hasConnectedOnce {
                self.listener.onReconnected()
            } else {
                self.hasConnectedOnce = true
                self.listener.onConnected()
            }
        }
        s.on(clientEvent: .disconnect) { [weak self] _, _ in
            self?.listener.onDisconnected()
        }

        s.on("peer-joined") { [weak self] data, _ in
            guard let obj = data.first as? [String: Any],
                  let peerId = obj["peerId"] as? String else { return }
            self?.listener.onPeerJoined(peerId)
        }
        s.on("peer-left") { [weak self] data, _ in
            guard let obj = data.first as? [String: Any],
                  let peerId = obj["peerId"] as? String else { return }
            self?.listener.onPeerLeft(peerId)
        }
        s.on("signal") { [weak self] data, _ in
            guard let obj = data.first as? [String: Any],
                  let from = obj["from"] as? String,
                  let payload = obj["data"] as? [String: Any] else { return }
            self?.listener.onSignal(from, payload)
        }

        manager = mgr
        socket = s
        s.connect()
    }

    /// Join a room. The ack carries the existing peers (empty when first in).
    /// - Parameter onAck: invoked with (ok, peers, reason).
    func joinRoom(_ roomId: String,
                  onAck: @escaping (_ ok: Bool, _ peers: [String], _ reason: String?) -> Void) {
        socket?.emitWithAck("join-room", ["roomId": roomId]).timingOut(after: 10) { response in
            guard let obj = response.first as? [String: Any] else {
                onAck(false, [], "no-response")
                return
            }
            if (obj["ok"] as? Bool) == true {
                let peers = (obj["peers"] as? [Any])?.compactMap { $0 as? String } ?? []
                onAck(true, peers, nil)
            } else {
                onAck(false, [], (obj["reason"] as? String) ?? "unknown")
            }
        }
    }

    func sendSignal(to: String, data: [String: Any]) {
        socket?.emit("signal", ["to": to, "data": data])
    }

    func leaveRoom(_ roomId: String) {
        socket?.emit("leave-room", ["roomId": roomId])
    }

    func disconnect() {
        socket?.removeAllHandlers()
        socket?.disconnect()
        socket = nil
        manager = nil
        hasConnectedOnce = false
    }
}
